### PR TITLE
Use setup_requires to ensure cython and numpy are present

### PR DESCRIPTION
**Pull request by j-santander**
Addressing issue #256 
Summary of changes:

I'm using setup_requires to indicate the dependencies on "setup" time, with install_requires to indicate the dependencies on "run" time. 

I'm also ensuring the setup_requires is evaluated before making any reference to them.

As numpy version I've used the declared dependency for Pandas 0.17.0

### CODE REVIEW

**Comment by wesm** [comment]
Need to make sure this doesn't conflict with conda environments -- setuptools has a way of wanting to build NumPy or Cython from source when it's already installed in your site-packages

**Comment by tdsmith** [comment]
A problem with using setup_requires this way is that, if they're missing, setuptools will install a temporary copy of cython and numpy for each setup.py invocation. Unfortunately, tools like pip will call setup.py several times to do things like extract metadata.

One solution is to cargo-cult this function around: https://github.com/pyca/cryptography/blob/8e66ca6813016d9fc6f57d5f1e50530fc39f78ae/setup.py#L114 -- the cryptography project runs into a similar chicken-egg problem with cffi.

Mercifully, this will be resolved in pip 10 by [PEP 518](https://www.python.org/dev/peps/pep-0518/) (pyproject.toml).

pip PEP 518 tracking bug: https://github.com/pypa/pip/issues/3691

**Comment by tdsmith** [comment]
-- although that doesn't model a complete solution for feather since cryptography manages to avoid actually importing cffi directly, relying on cffi's setuptools extension hooks instead; packaging is hard.

**Comment by j-santander** [comment]
Yes, I know :) Thanks for looking into this.

**Comment by danpalmer** [comment]
Ping on this? I just tried to install `feather-format` and it failed. Not a great first-run experience for someone new to the format.

It also means we can't use a `requirements.txt`, as pip doesn't know the dependency and therefore can't install `numpy` and `Cython` first.

File: python/setup.py [hunk 1]
```diff
@@ -29,6 +25,18 @@
 import os
 import platform
 
+#
+# Force `setup_requires` stuff like Cython to be installed before proceeding
+#
+from setuptools.dist import Distribution
+Distribution(dict(setup_requires=['cython>=0.21', 'numpy>=1.7.1']))
```
Comment (danpalmer) [inline python/setup.py]: The correct name of the package is `Cython` – using the lowercase version will result in issues with some tooling as it is not the canonical name.

**Comment by wesm** [comment]
Closing as Cython is no longer required to create the feather Python package (which is just a convenience interface to `pyarrow.feather`)

**Pull request closed without merging**